tutorial/developing_packages/editable_cmake_packages/say/conanfile.py

```diff
@@ -0,0 +1,46 @@
+from conan import ConanFile
+from conan.tools.cmake import CMakeToolchain, CMake, cmake_layout
+
+
+class SayConan(ConanFile):
+    name = "say"
+    version = "1.0"
+
+    # Binary configuration
+    settings = "os", "compiler", "build_type", "arch"
+    options = {"shared": [True, False], "fPIC": [True, False], "something_custom": [True, False]}
+    default_options = {"shared": False, "fPIC": True, "something_custom": False}
+
+    # Sources are located in the same place as this recipe, copy them to the recipe
+    exports_sources = "CMakeLists.txt", "src/*", "include/*"
+
+    def config_options(self):
+        if self.settings.os == "Windows":
+            del self.options.fPIC
+
+    def validate(self):
+        # ensure the consumer's options affect this dependency
+        if not self.options.something_custom:
+            raise ConanInvalidConfiguration("The consumer should set something_custom=True")
+
+    def layout(self):
+        cmake_layout(self)
+
+        self.cpp.source.includedirs = ["thelib/src"]
+        self.cpp.build.libdirs = ["thelib/src"]
```

Comment:
This doesn't seem a valid ``cpp.build.libdirs``, I don't think the final ``say.lib`` goes in ``thelib/src/say.lib``.

Reply:
It sure does.
```
$ ls build_say/build/Release/thelib/src/
CMakeFiles  cmake_install.cmake  libsay.a
```

This matches the way a real library is built.
The source is not at the top level, there are several directory levels with documentation and other things before it gets to the source, and that is where cmake builds the library.

A key thing that is demonstrated here is the libraries will end up in a very different path to after it is packaged.
ie after packaging, it is in the sensible `package/lib` folder.  Before, libraries may be getting built in all sorts of places.

Reply:
@paulharris keep in mind that people use different platforms and generators. cmake does do different build folder layouts depending on that.

i've been using the following expression usually:
```py
# this sets self.cpp.build.libdirs[0] (and bindirs) to a reasonable value based on build context
# it usually becomes "." or self.build_type
cmake_layout(self)

# example below expects something like
# /CMakeLists.txt -> add_subdirectory(relative/path/to/add_library/call)
# /relative/path/to/add_library/call/CMakeLists.txt -> add_library(foo)
self.cpp.build.libdirs = [
  os.path.join("relative", "path", "to", "add_library", "call", self.cpp.build.libdirs[0])
]
self.cpp.build.bindirs = [
  os.path.join("relative", "path", "to", "add_library", "call", self.cpp.build.bindirs[0])
]
self.cpp.build.libs = ["foo"]
```